File: chatguard-web/src/lib/crypto/__tests__/groupKeyManager.test.ts
import { describe, test, expect, beforeEach, vi } from 'vitest';
import { GroupKeyManager } from '../groupKeyManager';
import { useGroupStore, type Group, type GroupKey } from '../../storage/groupStore';

// Mock Shield
vi.mock('../../shield/crypto', () => ({
  Shield: {
    randomBytes: (size: number) => new Uint8Array(size).fill(0x42),
    quickEncrypt: (key: Uint8Array, data: Uint8Array) => new Uint8Array([...data, 0xee]),
    quickDecrypt: (key: Uint8Array, data: Uint8Array) => data.slice(0, -1),
  },
  shieldCrypto: {
    encryptMessage: vi.fn().mockResolvedValue(new Uint8Array([0x01, 0x02, 0x03])),
    decryptMessage: vi.fn().mockResolvedValue('dGVzdA=='), // base64 for 'test'
  },
}));

// Mock crypto.randomUUID
vi.stubGlobal('crypto', {
  randomUUID: () => 'test-uuid-' + Math.random().toString(36).substring(7),
});

describe('GroupKeyManager', () => {
  let keyManager: GroupKeyManager;

  beforeEach(() => {
    keyManager = new GroupKeyManager();
    // Reset group store
    useGroupStore.setState({ groups: [], members: {}, messages: {}, keys: [] });
  });

  describe('generateGroupKey', () => {
    test('creates 32-byte key', () => {
      const key = keyManager.generateGroupKey();
      expect(key.length).toBe(32);
    });

    test('generates random bytes', () => {
      const key = keyManager.generateGroupKey();
      expect(key[0]).toBe(0x42); // Our mock fills with 0x42
    });
  });

  describe('generateKeyId', () => {
    test('creates unique ID', () => {
      const id1 = keyManager.generateKeyId();
      const id2 = keyManager.generateKeyId();
      expect(id1).not.toBe(id2);
    });

    test('ID is UUID format', () => {
      const id = keyManager.generateKeyId();
      expect(id).toContain('test-uuid-');
    });
  });

  describe('createGroupKey', () => {
    test('creates and stores group key', async () => {
      const groupId = 'test-group-1';
      const groupKey = await keyManager.createGroupKey(groupId);

      expect(groupKey.groupId).toBe(groupId);
      expect(groupKey.rotationNumber).toBe(0);
      expect(groupKey.encryptedKey.length).toBeGreaterThan(0);

      // Verify stored in group store
      const stored = useGroupStore.getState().keys;
      expect(stored.length).toBe(1);
      expect(stored[0].id).toBe(groupKey.id);
    });
  });

  describe('rotateKey', () => {
    test('generates new key with incremented rotation', async () => {
      const group: Group = {
        id: 'test-group-1',
        name: 'Test Group',
        createdAt: Date.now(),
        createdBy: '',
        currentKeyId: 'old-key-id',
        keyRotationCount: 2,
      };

      const newKey = await keyManager.rotateKey(group);

      expect(newKey.rotationNumber).toBe(3); // 2 + 1
      expect(newKey.id).not.toBe('old-key-id');
    });

    test('stores new key in store', async () => {
      const group: Group = {
        id: 'test-group-1',
        name: 'Test Group',
        createdAt: Date.now(),
        createdBy: '',
        currentKeyId: 'old-key-id',
        keyRotationCount: 0,
      };

      const newKey = await keyManager.rotateKey(group);

      const stored = useGroupStore.getState().keys;
      expect(stored.some((k) => k.id === newKey.id)).toBe(true);
    });
  });

  describe('encryptGroupMessage', () => {
    test('encrypts with current group key', async () => {
      const groupId = 'test-group-1';

      // First create a key for the group
      const groupKey = await keyManager.createGroupKey(groupId);

      // Create the group
      useGroupStore.getState().addGroup({
        id: groupId,
        name: 'Test',
        createdAt: Date.now(),
        createdBy: '',
        currentKeyId: groupKey.id,
        keyRotationCount: 0,
      });

      const plaintext = new Uint8Array([0x01, 0x02, 0x03]);
      const encrypted = await keyManager.encryptGroupMessage(groupId, plaintext);

      // Our mock adds 0xee at the end
      expect(encrypted[encrypted.length - 1]).toBe(0xee);
    });

    test('throws error when no key exists', async () => {
      const groupId = 'nonexistent-group';

      await expect(
        keyManager.encryptGroupMessage(groupId, new Uint8Array([1]))
      ).rejects.toThrow();
    });
  });

  describe('decryptGroupMessage', () => {
    test('decrypts with specified key', async () => {
      const groupId = 'test-group-1';

      // Create and store a key
      const groupKey = await keyManager.createGroupKey(groupId);

      useGroupStore.getState().addGroup({
        id: groupId,
        name: 'Test',
        createdAt: Date.now(),
        createdBy: '',
        currentKeyId: groupKey.id,
        keyRotationCount: 0,
      });

      // Mock ciphertext (plaintext + 0xee from mock)
      const ciphertext = new Uint8Array([0x01, 0x02, 0x03, 0xee]);

      const decrypted = await keyManager.decryptGroupMessage(groupId, groupKey.id, ciphertext);

      // Our mock removes the last byte
      expect(Array.from(decrypted)).toEqual([0x01, 0x02, 0x03]);
    });

    test('throws error for unknown key', async () => {
      await expect(
        keyManager.decryptGroupMessage('group-1', 'unknown-key', new Uint8Array([1]))
      ).rejects.toThrow();
    });
  });

  describe('storeReceivedKey', () => {
    test('stores key from another member', async () => {
      const groupId = 'test-group-1';
      const keyId = 'received-key-id';
      const keyBytes = new Uint8Array(32).fill(0x11);
      const rotationNumber = 5;

      await keyManager.storeReceivedKey(groupId, keyId, keyBytes, rotationNumber);

      const stored = useGroupStore.getState().keys;
      const key = stored.find((k) => k.id === keyId);
      expect(key).toBeDefined();
      expect(key?.rotationNumber).toBe(5);
      expect(key?.groupId).toBe(groupId);
    });
  });

  describe('hasKeyForGroup', () => {
    test('returns false when no key exists', () => {
      expect(keyManager.hasKeyForGroup('unknown-group')).toBe(false);
    });

    test('returns true when key exists', async () => {
      const groupId = 'test-group-1';
      const groupKey = await keyManager.createGroupKey(groupId);

      useGroupStore.getState().addGroup({
        id: groupId,
        name: 'Test',
        createdAt: Date.now(),
        createdBy: '',
        currentKeyId: groupKey.id,
        keyRotationCount: 0,
      });

      expect(keyManager.hasKeyForGroup(groupId)).toBe(true);
    });
  });
});

describe('GroupMessageEnvelope', () => {
  test('serialize and deserialize preserves all fields', () => {
    const envelope = {
      type: 'TEXT' as const,
      groupId: 'group-1',
      senderId: 'sender-1',
      messageId: 'msg-1',
      timestamp: 1705780000000,
      keyId: 'key-1',
      content: 'Hello group!',
    };

    const json = JSON.stringify(envelope);
    const restored = JSON.parse(json);

    expect(restored).toEqual(envelope);
  });

  test('MEMBER_ADDED includes member contact ID', () => {
    const envelope = {
      type: 'MEMBER_ADDED' as const,
      groupId: 'group-1',
      senderId: 'admin-1',
      messageId: 'msg-1',
      timestamp: Date.now(),
      keyId: 'key-1',
      metadata: { memberId: 'new-member-1' },
    };

    const json = JSON.stringify(envelope);
    expect(json).toContain('new-member-1');
  });

  test('KEY_ROTATION includes new key ID', () => {
    const envelope = {
      type: 'KEY_ROTATION' as const,
      groupId: 'group-1',
      senderId: 'admin-1',
      messageId: 'msg-1',
      timestamp: Date.now(),
      keyId: 'new-key-123',
      metadata: { oldKeyId: 'old-key-456' },
    };

    const json = JSON.stringify(envelope);
    expect(json).toContain('new-key-123');
    expect(json).toContain('old-key-456');
  });
});

describe('Group Permission Tests', () => {
  test('admin role check returns correct value', () => {
    const members = [
      { groupId: 'g1', contactId: '', role: 'admin' as const, displayName: 'You', joinedAt: 0, addedBy: '' },
      { groupId: 'g1', contactId: 'c1', role: 'member' as const, displayName: 'Alice', joinedAt: 0, addedBy: '' },
    ];

    useGroupStore.setState({ members: { g1: members }, groups: [], messages: {}, keys: [] });

    const selfMember = useGroupStore.getState().members['g1']?.find((m) => m.contactId === '');
    expect(selfMember?.role).toBe('admin');
  });

  test('non-admin role check returns correct value', () => {
    const members = [
      { groupId: 'g1', contactId: '', role: 'member' as const, displayName: 'You', joinedAt: 0, addedBy: 'admin' },
      { groupId: 'g1', contactId: 'admin', role: 'admin' as const, displayName: 'Admin', joinedAt: 0, addedBy: '' },
    ];

    useGroupStore.setState({ members: { g1: members }, groups: [], messages: {}, keys: [] });

    const selfMember = useGroupStore.getState().members['g1']?.find((m) => m.contactId === '');
    expect(selfMember?.role).toBe('member');
  });
});

File: chatguard-android/app/src/test/java/ai/guard8/chatguard/crypto/GroupKeyManagerTest.kt
package ai.guard8.chatguard.crypto

import ai.guard8.chatguard.model.Group
import ai.guard8.chatguard.model.GroupKey
import ai.guard8.chatguard.model.GroupMember
import ai.guard8.chatguard.model.MemberRole
import ai.guard8.chatguard.storage.GroupDao
import io.mockk.*
import kotlinx.coroutines.runBlocking
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

/**
 * Unit tests for GroupKeyManager.
 * Tests key generation, rotation, encryption, and distribution.
 */
class GroupKeyManagerTest {

    private lateinit var keyManager: KeyManager
    private lateinit var shieldCrypto: ShieldCrypto
    private lateinit var groupDao: GroupDao
    private lateinit var groupKeyManager: GroupKeyManager

    @Before
    fun setup() {
        keyManager = mockk()
        shieldCrypto = mockk()
        groupDao = mockk()

        // Mock master key retrieval
        every { keyManager.retrieveKey("group_master_key") } returns ByteArray(32) { 0x42 }

        groupKeyManager = GroupKeyManager(keyManager, shieldCrypto, groupDao)
    }

    // ==================== Key Generation Tests ====================

    @Test
    fun `generateGroupKey creates 32-byte key`() {
        val key = groupKeyManager.generateGroupKey()
        assertEquals(32, key.size)
    }

    @Test
    fun `generateGroupKey creates random keys`() {
        val key1 = groupKeyManager.generateGroupKey()
        val key2 = groupKeyManager.generateGroupKey()

        // Keys should be different (cryptographically random)
        assertFalse("Keys should be random", key1.contentEquals(key2))
    }

    @Test
    fun `generateKeyId creates unique UUIDs`() {
        val id1 = groupKeyManager.generateKeyId()
        val id2 = groupKeyManager.generateKeyId()

        assertNotEquals("Key IDs should be unique", id1, id2)
        // UUID format check
        assertTrue("Should be UUID format", id1.matches(Regex("[a-f0-9-]{36}")))
    }

    // ==================== Key Creation Tests ====================

    @Test
    fun `createGroupKey creates key with rotation number 0`() = runBlocking {
        val groupId = "test-group-1"

        val groupKey = groupKeyManager.createGroupKey(groupId)

        assertEquals(groupId, groupKey.groupId)
        assertEquals(0, groupKey.rotationNumber)
        assertNotNull(groupKey.id)
        assertTrue(groupKey.encryptedKey.isNotEmpty())
    }

    // ==================== Key Rotation Tests ====================

    @Test
    fun `rotateKey generates new key with incremented rotation number`() = runBlocking {
        val group = Group(
            id = "test-group-1",
            name = "Test Group",
            createdAt = System.currentTimeMillis(),
            createdBy = "",
            currentKeyId = "old-key-id",
            keyRotationCount = 2
        )

        coEvery { groupDao.deleteOldKeys(any(), any()) } just Runs

        val newKey = groupKeyManager.rotateKey(group)

        assertEquals(group.id, newKey.groupId)
        assertEquals(3, newKey.rotationNumber) // 2 + 1
        assertNotEquals("old-key-id", newKey.id)
    }

    @Test
    fun `rotateKey cleans up old keys after MAX_OLD_KEYS_TO_KEEP`() = runBlocking {
        val group = Group(
            id = "test-group-1",
            name = "Test",
            createdAt = System.currentTimeMillis(),
            createdBy = "",
            currentKeyId = "old-key",
            keyRotationCount = 15 // After rotation: 16, should delete keys < 7
        )

        coEvery { groupDao.deleteOldKeys(any(), any()) } just Runs

        groupKeyManager.rotateKey(group)

        coVerify { groupDao.deleteOldKeys("test-group-1", 7) }
    }

    @Test
    fun `rotateKey does not delete keys when under threshold`() = runBlocking {
        val group = Group(
            id = "test-group-1",
            name = "Test",
            createdAt = System.currentTimeMillis(),
            createdBy = "",
            currentKeyId = "old-key",
            keyRotationCount = 3 // After rotation: 4, threshold would be -5 (negative)
        )

        groupKeyManager.rotateKey(group)

        // Should not call deleteOldKeys when threshold is <= 0
        coVerify(exactly = 0) { groupDao.deleteOldKeys(any(), any()) }
    }

    // ==================== Encryption Tests ====================

    @Test
    fun `encryptKeyForMember uses ShieldCrypto`() {
        val groupKey = ByteArray(32) { it.toByte() }
        val contactId = "contact-1"
        val expectedCiphertext = ByteArray(50) { 0xEE.toByte() }

        every { shieldCrypto.encryptMessage(contactId, true, any()) } returns expectedCiphertext

        val encrypted = groupKeyManager.encryptKeyForMember(groupKey, contactId, true)

        verify { shieldCrypto.encryptMessage(contactId, true, any()) }
        assertEquals(expectedCiphertext, encrypted)
    }

    @Test
    fun `decryptKeyFromMember returns decrypted key`() {
        val encryptedKey = ByteArray(50) { 0xEE.toByte() }
        val contactId = "contact-1"
        val base64Key = "AQIDBAUG" // Some base64

        every { shieldCrypto.decryptMessage(contactId, false, encryptedKey) } returns base64Key

        val decrypted = groupKeyManager.decryptKeyFromMember(encryptedKey, contactId, false)

        verify { shieldCrypto.decryptMessage(contactId, false, encryptedKey) }
        assertNotNull(decrypted)
    }

    // ==================== Group Message Encryption Tests ====================

    @Test
    fun `encryptGroupMessage throws when no key exists`() = runBlocking {
        val groupId = "nonexistent-group"

        coEvery { groupDao.getCurrentKey(groupId) } returns null

        try {
            groupKeyManager.encryptGroupMessage(groupId, "test".toByteArray())
            fail("Should throw exception when no key exists")
        } catch (e: IllegalStateException) {
            assertTrue(e.message?.contains("No group key found") == true)
        }
    }

    @Test
    fun `decryptGroupMessage throws for unknown key`() = runBlocking {
        val groupId = "test-group"
        val keyId = "unknown-key"

        coEvery { groupDao.getKeyById(keyId) } returns null

        try {
            groupKeyManager.decryptGroupMessage(groupId, keyId, ByteArray(32))
            fail("Should throw exception for unknown key")
        } catch (e: IllegalStateException) {
            assertTrue(e.message?.contains("Group key not found") == true)
        }
    }

    // ==================== Key Storage Tests ====================

    @Test
    fun `storeReceivedKey stores key in DAO`() = runBlocking {
        val groupId = "test-group-1"
        val keyId = "received-key-id"
        val keyBytes = ByteArray(32) { 0x11 }
        val rotationNumber = 5

        coEvery { groupDao.insertKey(any()) } just Runs

        groupKeyManager.storeReceivedKey(groupId, keyId, keyBytes, rotationNumber)

        coVerify {
            groupDao.insertKey(match {
                it.id == keyId &&
                it.groupId == groupId &&
                it.rotationNumber == rotationNumber
            })
        }
    }

    // ==================== Has Key Tests ====================

    @Test
    fun `hasKeyForGroup returns false when no key exists`() = runBlocking {
        val groupId = "unknown-group"

        coEvery { groupDao.getCurrentKey(groupId) } returns null

        assertFalse(groupKeyManager.hasKeyForGroup(groupId))
    }

    @Test
    fun `hasKeyForGroup returns true when key exists`() = runBlocking {
        val groupId = "test-group"
        val groupKey = GroupKey(
            id = "k1",
            groupId = groupId,
            encryptedKey = ByteArray(32),
            createdAt = System.currentTimeMillis(),
            rotationNumber = 0
        )

        coEvery { groupDao.getCurrentKey(groupId) } returns groupKey

        assertTrue(groupKeyManager.hasKeyForGroup(groupId))
    }

    // ==================== Key Distribution Tests ====================

    @Test
    fun `prepareKeyDistribution encrypts for all non-self members`() = runBlocking {
        val groupId = "test-group-1"
        val groupKey = ByteArray(32) { it.toByte() }
        val members = listOf(
            GroupMember(groupId, "", "Me", MemberRole.ADMIN, System.currentTimeMillis(), ""),
            GroupMember(groupId, "c1", "Alice", MemberRole.MEMBER, System.currentTimeMillis(), ""),
            GroupMember(groupId, "c2", "Bob", MemberRole.MEMBER, System.currentTimeMillis(), "")
        )

        val encryptedForAlice = ByteArray(50) { 0xAA.toByte() }
        val encryptedForBob = ByteArray(50) { 0xBB.toByte() }

        every { shieldCrypto.encryptMessage("c1", true, any()) } returns encryptedForAlice
        every { shieldCrypto.encryptMessage("c2", true, any()) } returns encryptedForBob

        val distribution = groupKeyManager.prepareKeyDistribution(groupId, members, groupKey)

        assertEquals(2, distribution.size) // Only Alice and Bob, not self
        assertEquals(encryptedForAlice, distribution["c1"])
        assertEquals(encryptedForBob, distribution["c2"])
    }

    @Test
    fun `prepareKeyDistribution skips self member`() = runBlocking {
        val groupId = "test-group-1"
        val groupKey = ByteArray(32)
        val members = listOf(
            GroupMember(groupId, "", "Me", MemberRole.ADMIN, System.currentTimeMillis(), "")
        )

        val distribution = groupKeyManager.prepareKeyDistribution(groupId, members, groupKey)

        assertTrue("Should not include self", distribution.isEmpty())
    }
}
